package-power-bin/main.qml

```diff
@@ -312,30 +306,26 @@ Kirigami.ApplicationWindow {
                     fanSlider.visible = false
                 }
                 fanTimer.running = true
-                disconnectSource(sourceName)
             }
         }
 
-        // Checks the current fan profile
-        PlasmaCore.DataSource {
-            id: fanProfilesChecker
-            engine: "executable"
-            connectedSources: []
-            onNewData: {
-                if (data["stdout"].trim() !== '') {
-                    fanSlider.value = fanProfilesModel.profileNames.indexOf(data["stdout"].trim())
+        ShellEngine {
+            id: altFanProfilesChecker
+            onStdoutChanged: {
+                if (stdout.trim() !== '') {
```

Comment:
Trim once and store? 
```javascript
// If you need to do more with trimmed_line, it will propagate everywhere it is used.
trimmed_line = stdout.trim();
if ( trimmed_line !=== '' ){ ... }
```

Reply:
I'm not sure if that's helpful. There is stdout.trim() all throughout the code, but keep in mind that these are in signal handlers and that the stdout string being accessed each time is different. I don't think there's anywhere else in the code that needs to know what the current fan profile is.

Reply:
(for instance, the string being referenced by stdout.trim() in altProfilesChecker is an entirely different string than the one referenced by stdout.trim() in altFanProfilesChecker.)

Reply:
If a value is computed and used more than once, then many think it should be assigned to a variable so that any changes in the computation will correctly propagate to all uses. For example, if the computed value were to be modified to use ucfirst(), then that would be reflected in the two later uses.

With just two uses, it's not something I'm going to lose sleep over. But if it were three, I'd be giving you a hard time! :)

Reply:
@mmikowski I think the problem is you're misunderstanding what's happening in the code.

The two places where stdout.trim() is called, two entirely different shell scripts are being called to do two different things that give different standard output that are used in different widgets. Neither widget can use the output of the other script.

The stdout property *is not* the same thing in each time. Think about it like this - in one spot, stdout is used in the altFanProfilesChecker object, and in another place, it's used in the altProfilesChecker object. So it's not *really* stdout both times - in one spot, it's altFanProfilesChecker.stdout, and in another spot it's altProfilesChecker.stdout. Those are different, and the only reason they share a name is because ShellEngine happens to provide a script's standard output in the stdout property. It's local to the object.

Were I to take the output of stdout.trim() and place it into a local variable, then use the local variable everywhere where stdout is used, the code would occasionally attempt to determine the active fan profile from the output of `kfocus-power-set -r`, and it would try to figure out the active power profile from the output of `kfocus-fan-set -r`. Clearly that's incorrect :)

The reason I can just reference "stdout" each time rather than using the full name "altProfilesChecker.stdout" is because I'm inside an event handler that is also local to the corresponding ShellEngine object. So my use of "stdout" is more like "this.stdout", which would explicitly reference the local object. I just don't need that explicit reference since I'm already "inside" one of the objects by virtue of being in an event handler attached to that object, so the fact that "stdout" is referring to the *local* stdout can be inferred.

Reply:
@mmikowski :facepalm: I just now figured out what you're getting at. I thought you were wanting me to make a trimmed_line variable that was global to the entire app, not local to the object itself. Never mind, I can fix that.

Reply:
Alright, cleaned up.

Reply:
I definitely understand lexical scope in functions. Here is one rewritten to use a variable and avoid trimming two additional times:

```javascript
           onStdoutChanged: {
                const trim_str = stdout.trim()
                if (trim_str !== '' && trim_str.substring(0, 6).toLowerCase() != "custom") {
                    profilesModel.selectedProfile = trim_str
                }
                disconnectSource(sourceName)
            }
```
To me, it's wasteful and confusing to trim 3 times. I'm pretty sure my published JS code style guide recommends this.

Reply:
That's too funny. I went to revisit the code and you had already changed it!